jlm/rvsdg/bitstring/arithmetic.cpp

```diff
@@ -1,115 +1,240 @@
-  std::unique_ptr<operation> NAME##_op::copy() const                     \
-  {                                                                      \
-    return std::unique_ptr<operation>(new NAME##_op(*this));             \
-  }                                                                      \
-                                                                         \
-  std::unique_ptr<bitbinary_op> NAME##_op::create(size_t nbits) const    \
-  {                                                                      \
-    return std::unique_ptr<bitbinary_op>(new NAME##_op(nbits));          \
-  }
-
-DEFINE_BITUNARY_OPERATION(bitneg, arg.neg(), BITNEGATE)
-DEFINE_BITUNARY_OPERATION(bitnot, arg.lnot(), BITNOT)
-
-DEFINE_BITBINARY_OPERATION(
-    bitadd,
-    arg1.add(arg2),
-    BITADD,
-    binary_op::flags::associative | binary_op::flags::commutative)
-DEFINE_BITBINARY_OPERATION(bitand, arg1.land(arg2), BITAND, binary_op::flags::associative | binary_op::flags::commutative)
-DEFINE_BITBINARY_OPERATION(bitashr, arg1.ashr(arg2.to_uint()), BITASHR, binary_op::flags::none)
-DEFINE_BITBINARY_OPERATION(
-    bitmul,
-    arg1.mul(arg2),
-    BITMUL,
-    binary_op::flags::associative | binary_op::flags::commutative)
-DEFINE_BITBINARY_OPERATION(
-    bitor,
-    arg1.lor(arg2),
-    BITOR,
-    binary_op::flags::associative | binary_op::flags::commutative)
-DEFINE_BITBINARY_OPERATION(bitsdiv, arg1.sdiv(arg2), BITSDIV, binary_op::flags::none)
-DEFINE_BITBINARY_OPERATION(bitshl, arg1.shl(arg2.to_uint()), BITSHL, binary_op::flags::none)
-DEFINE_BITBINARY_OPERATION(bitshr, arg1.shr(arg2.to_uint()), BITSHR, binary_op::flags::none)
-DEFINE_BITBINARY_OPERATION(bitsmod, arg1.smod(arg2), BITSMOD, binary_op::flags::none)
-DEFINE_BITBINARY_OPERATION(bitsmulh, arg1.smulh(arg2), BITSMULH, binary_op::flags::none)
-DEFINE_BITBINARY_OPERATION(bitsub, arg1.sub(arg2), BITSUB, binary_op::flags::none)
-DEFINE_BITBINARY_OPERATION(bitudiv, arg1.udiv(arg2), BITUDIV, binary_op::flags::none)
-DEFINE_BITBINARY_OPERATION(bitumod, arg1.umod(arg2), BITUMOD, binary_op::flags::none)
-DEFINE_BITBINARY_OPERATION(bitumulh, arg1.umulh(arg2), BITUMULH, binary_op::flags::none)
-DEFINE_BITBINARY_OPERATION(
-    bitxor,
-    arg1.lxor(arg2),
-    BITXOR,
-    binary_op::flags::associative | binary_op::flags::commutative)
+// This provides the explicit template instantiations underlying
+// all bitstring operation classes: This ensures that there is a
+// single definition for all virtual functions, vmts etc in the
+// rvsdg library (instead of the compiler template-instantiating
+// them multiple times in each translation unit, and then relying
+// on linker to do de-duplication).
+
+struct reduce_neg
+{
+  bitvalue_repr
+  operator()(const bitvalue_repr & arg) const
+  {
+    return arg.neg();
+  }
+};
+
+struct reduce_not
+{
+  bitvalue_repr
+  operator()(const bitvalue_repr & arg) const
+  {
+    return arg.lnot();
+  }
+};
+
+const char BITNEGATE_LABEL[] = "BITNEGATE";
```

Comment:
I think we can also move away from the capital letters as debug strings. `BitNegate` would just do nicely as well.

Reply:
Renamed, and also renamed the constant names themselves -- not sure if you did mean that as well?